### ISSUE TITLE
Support for alert(), confirm() and prompt() boxes?

### ISSUE THREAD

**Issue by sigmaris** [minkphp/Mink#158]
Is there any way using the Mink API to, for example, click on the OK or Cancel buttons on a Javascript-generated confirm dialog box or fill in text in a dialog generated from prompt()? We're having trouble testing steps that require the user to confirm certain actions such as deleting objects, since the confirmation doesn't seem to succeed. FWIW we're using Mink 1.3.2 with the ZombieDriver.

**Comment by J7mbo** [comment]
Any support for alerts yet? ;-)

**Comment by aik099** [comment]
Nope. There can't any support until a decision is made in this task how it can be implemented in a cross-driver fashion.

**Comment by stof** [comment]
@J7mbo no, because nobody has ever suggested an API which can be implemented across drivers. Until then, you will have to resort to driver-specific implementations (drivers generally expose a way to use their internal implementation. for instance, the Selenium2Driver has a ``getWebdriverSession`` method)

**Comment by J7mbo** [comment]
I understand, fair enough guys. Would be good if we could come up with a good solution for that sometime.

For the record if there's anyone else here wondering the same thing and just wants to 'get it working', here's what I'm doing in one of my steps for Selenium2Driver as Stof suggested:

        /** @var \Behat\Mink\Driver\Selenium2Driver $driver Needed because no cross-driver way yet */
        $driver = $this->getSession()->getDriver();

        /** Get the alert text (for your assertions) **/
        $driverMessage = $driver->getWebDriverSession()->getAlert_text();

        /** Accept the alert so you don't get a runtime exception when the steps try to continue **/
        $driver->getWebDriverSession()->accept_alert();

**Comment by supriyarajgopal-spi** [comment]
@J7mbo 's solution worked for me, thanks so much!

**Comment by mihaeu** [comment]
This approach did not work for me. I solved the problem by injecting JavaScript using the driver:

```php
    /**
     * @Given /^I confirm the dialog$/
     * @throws \Behat\Mink\Exception\DriverException
     */
    public function iConfirmTheDialog()
    {
        $this->getSession()->getDriver()->executeScript('window.confirm = function () { return true; };');
    }
```

This has to be called before triggering the confirm.

**Comment by tractorcow** [comment]
I suggest that instead of baking alert support directly into the DriverInterface (and thus immediately rendering all drivers inoperable), might I suggest an "optional" interface, which may be triggered for detection via capabilities?

Pseudocode follows:

```
interface DriverInterface {}
interface AlertSupportedDriverInterface extends DriverInterface {}
```

```
function buildDriver(array $config) {
	$class = $this->getClass();
	if (isset($config['capabilities']['alerts'])) {
		if (! is_a($class, AlertSupportedDriverInterface::class, true)) {
			throw new NotSuppportedCapabilityException('Alert capability not supported by this driver');
		}
	}
	// ...
}
```

Once the majority of drivers have been upgraded, this feature could be promoted to a standard DriverInterface.

**Comment by shwetaneelsharma** [comment]
@mihaeu You saved my day. Thanks a ton. Your solution worked for me on behat 3.*

**Comment by mihaeu** [comment]
> @mihaeu You saved my day. Thanks a ton. Your solution worked for me on behat 3.*

You're most welcome.

**Comment by ihsanfaisal** [comment]
> @J7mbo 's approach did not work for me. I solved the problem by injecting JavaScript using the driver:
> 
> ```
>     /**
>      * @Given /^I confirm the dialog$/
>      * @throws \Behat\Mink\Exception\DriverException
>      */
>     public function iConfirmTheDialog()
>     {
>         $this->getSession()->getDriver()->executeScript('window.confirm = function () { return true; };');
>     }
> ```
> 
> This has to be called before triggering the confirm.

Thank you for the solution. @mihaeu Would you care to explain how is injecting the script can solve the problem?
